libp2p/protocols/secure/noise.nim

```diff
@@ -0,0 +1,476 @@
+  # according to spec if key is empty leave plaintext
+  if ss.cs.hasKey:
+    result = ss.cs.encryptWithAd(ss.h.data, data)
+  else:
+    result = @data
+  ss.mixHash(result)
+
+proc decryptAndHash(ss: var SymmetricState, data: var openarray[byte]): seq[byte] =
+  # according to spec if key is empty leave plaintext
+  if ss.cs.hasKey:
+    result = ss.cs.decryptWithAd(ss.h.data, data)
+  else:
+    result = @data
+  ss.mixHash(data)
+
+proc split(ss: var SymmetricState): tuple[cs1, cs2: CipherState] =
+  var
+    temp_keys: array[2, ChaChaPolyKey]
+    nkeys = 0
+  for next in sha256.hkdf(ss.ck, [], []):
+    if nkeys > temp_keys.high:
+      break
+    temp_keys[nkeys] = next.data
+    inc nkeys
+
+  init result.cs1, temp_keys[0]
+  init result.cs2, temp_keys[1]
+  
+template write_e: untyped =
+  # Sets e (which must be empty) to GENERATE_KEYPAIR(). Appends e.public_key to the buffer. Calls MixHash(e.public_key).
+  hs.e = genKeyPair()
+  var ne = hs.e.publicKey
+  msg &= @ne
+  hs.ss.mixHash(ne)
+
+template write_s: untyped =
+  # Appends EncryptAndHash(s.public_key) to the buffer.
+  var spk = @(hs.s.publicKey)
+  msg &= hs.ss.encryptAndHash(spk)
+
+template dh_ee: untyped =
+  # Calls MixKey(DH(e, re)).
+  hs.ss.mixKey(dh(hs.e.privateKey, hs.re))
+
+template dh_es: untyped =
+  # Calls MixKey(DH(e, rs)) if initiator, MixKey(DH(s, re)) if responder.
+  when initiator:
+    hs.ss.mixKey(dh(hs.e.privateKey, hs.rs))
+  else:
+    hs.ss.mixKey(dh(hs.s.privateKey, hs.re))
+
+template dh_se: untyped =
+  # Calls MixKey(DH(s, re)) if initiator, MixKey(DH(e, rs)) if responder.
+  when initiator:
+    hs.ss.mixKey(dh(hs.s.privateKey, hs.re))
+  else:
+    hs.ss.mixKey(dh(hs.e.privateKey, hs.rs))
+
+template dh_ss: untyped =
+  # Calls MixKey(DH(s, rs)).
+  hs.ss.mixKey(dh(hs.s.privateKey, hs.rs))
+
+template read_e: untyped =
+  # Sets re (which must be empty) to the next DHLEN bytes from the message. Calls MixHash(re.public_key).
+  copyMem(addr hs.re[0], addr msg[0], Curve25519Key.len)
+  msg = msg[Curve25519Key.len..msg.high]
+  hs.ss.mixHash(hs.re)
+
+template read_s: untyped =
+  # Sets temp to the next DHLEN + 16 bytes of the message if HasKey() == True, or to the next DHLEN bytes otherwise.
+  # Sets rs (which must be empty) to DecryptAndHash(temp).
+  var temp: seq[byte]
+  if hs.ss.cs.hasKey:
+    temp.setLen(Curve25519Key.len + 16)
+    copyMem(addr temp[0], addr msg[0], Curve25519Key.len + 16)
+    msg = msg[Curve25519Key.len + 16..msg.high]
+  else:
+    temp.setLen(Curve25519Key.len)
+    copyMem(addr temp[0], addr msg[0], Curve25519Key.len)
+    msg = msg[Curve25519Key.len..msg.high]
+  var plain = hs.ss.decryptAndHash(temp)
+  copyMem(addr hs.rs[0], addr plain[0], Curve25519Key.len)
+  
+proc handshakeXXOutbound(p: Noise, conn: Connection, p2pProof: ProtoBuffer): Future[tuple[cs1, cs2: CipherState; remotePrologue: seq[byte]; rs: Curve25519Key]] {.async.} =
```

Comment:
Why not declare an object for handshake's result, it looks like it will be used as result at least in `handshakeXXOutbound` and `handshakeXXInbound` and when you add more types of handshakes i think this result type object will be reused there also.

Reply:
agree, this needs to be a type.

Reply:
Cool, done!